### PR TITLE
gh-90449: Improve accuracy and readability of exceptions tutorial

### PR DESCRIPTION
**Pull request by iritkatriel**
https://bugs.python.org/issue46291

### CODE REVIEW

File: Doc/tutorial/errors.rst
Comment (iritkatriel) [inline Doc/tutorial/errors.rst:150]: This paragraph needed to move up because the next example uses a named except clause. I also reworded it a bit.

File: Doc/tutorial/errors.rst
Comment (iritkatriel) [inline Doc/tutorial/errors.rst:347]: This is mentioned above as well, so I removed the repetition.

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]: Not a fan of this section!  `except Exception` is the catch-all recommended by style guides and linting tools; you have to know what you’re doing when you write `except BaseException`.

I saw this was discussed in the ticket but didn’t read all messages.  Do you think it’s in scope for this ticket, or should be a discussion somewhere with a wider core dev audience?

Comment (iritkatriel) [reply Doc/tutorial/errors.rst]: We can discuss it here I think. I agree it's a dodgy thing to have in a tutorial.  And we probably don't want to talk about the no-type except: clause here either.

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]:
```suggestion
:exc:`BaseException` is the common base class of all exceptions, including a few
```

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]:
```suggestion
program should terminate. They include :exc:`SystemExit` which is raised by
```

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]: Is `sys.exit` not a link on purpose?

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]: This sentence has has two verbs

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]: This repeats lines 180-184, should probably be adapted before merge (no rush!)

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]:
```suggestion
       print("OS error:", err)
```

I often see that people love f-strings and forget that `print` is amazing!  Accepts multiple params, converts them to strings, adds a space or something else between them, all that.

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst:208]: Personnally I find this formatting a little advanced for a tuto, but it was here before so it is what it is.

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]: I don’t think the previous use of `in` was incorrect or confusing, and don’t really like `from` here.

**Review by merwok — CHANGES_REQUESTED** [review]
This is shaping up nicely!

File: Doc/tutorial/errors.rst
Comment (iritkatriel) [inline Doc/tutorial/errors.rst]:
```suggestion
wishes to interrupt the program.
```

File: Doc/tutorial/errors.rst
Comment (iritkatriel) [inline Doc/tutorial/errors.rst]:
```suggestion
indirectly) in the *try clause*. For example::
```

File: Doc/tutorial/errors.rst
Comment (iritkatriel) [reply Doc/tutorial/errors.rst]: 180-184 are talking about BaseException, and this is about Exception.

File: Doc/tutorial/errors.rst
Comment (iritkatriel) [inline Doc/tutorial/errors.rst]:
```suggestion
:meth:`sys.exit` and :exc:`KeyboardInterrupt` which is raised when a user
```

**Comment by iritkatriel** [comment]
Apart from this one:
https://github.com/python/cpython/pull/31899/files#r833668489

I have made the requested changes; please review again.

File: Doc/tutorial/errors.rst
Comment (merwok) [reply Doc/tutorial/errors.rst]: Ah, I read that as an apposition `BaseException, which is a common etc`, so it is ambiguous.

Comment (iritkatriel) [reply Doc/tutorial/errors.rst]: Ah after re-reading I see the repetition, will fix!

File: Doc/tutorial/errors.rst
Comment (merwok) [inline Doc/tutorial/errors.rst]: Because we are in the tutorial, I wonder if we should say: `a class that derives from Exception or BaseException`.

Yes, BaseException implies Exception, but it may not be self-evident at this point, so this wording would be technically correct (exception = class deriving from BaseException) and friendly (you should remember Exception!)

Comment (iritkatriel) [reply Doc/tutorial/errors.rst]: Maybe 'a class that derive from BaseException, such as Exception or one if its subclasses' ?

File: Doc/tutorial/errors.rst
Comment (JelleZijlstra) [inline Doc/tutorial/errors.rst]: "apart" should be "apart from", but the whole sentence is rather complex. Here's a suggested rewrite:

```suggestion
:exc:`Exception` is the common base class of almost all exceptions. It is
itself a subclass of :exc:`BaseException`. A few other exceptions also
derive directly from :exc:`BaseException`. These should not typically be
handled, because they indicate that the program should terminate. They
include :exc:`SystemExit` which is raised by
```

This avoids overly long sentences and gives prominence to `Exception`, which I think is more important to cover in the tutorial.

Comment (iritkatriel) [reply Doc/tutorial/errors.rst]: I don't object to rewording this sentence, but I don't think the suggested change works.  This doesn't really make sense at this point:

"A few other exceptions also derive directly from :exc:`BaseException`, because they indicate that the program should terminate and should not typically be handled."

because it's not clear why the fact that "they indicate .. and should not be handled" explain why they are not BaseException subclasses.

Comment (JelleZijlstra) [reply Doc/tutorial/errors.rst]: Good point, reworded my suggestion above

Comment (iritkatriel) [reply Doc/tutorial/errors.rst]: That's better, but we are no longer saying that BaseException is the base class of all exceptions.

Comment (JelleZijlstra) [reply Doc/tutorial/errors.rst]: True, but do we need to? This is the tutorial, not the language reference.

Comment (iritkatriel) [reply Doc/tutorial/errors.rst]: I don't think it's clear why BaseException exists without that point.

Comment (iritkatriel) [reply Doc/tutorial/errors.rst]: I've reworded this paragraph, hopefully the new version is better.

**Comment by iritkatriel** [comment]
I have made the requested changes; please review again.

**Review by JelleZijlstra — APPROVED** [review]
Thanks, this is very readable

**Pull request merged**
